### PR TITLE
impr: update notification for c21 (FPLA-936)

### PR DESCRIPTION
**Pull request by Gogetter**
### Update notification for C21 [XS] ###

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java [hunk 1]
```diff
@@ -43,24 +48,34 @@ protected C21OrderEmailContentProvider(@Value("${ccd.ui.base.url}")String uiBase
...
         String hearingDate = "";
-        if (!ObjectUtils.isEmpty(caseData.getHearingDetails())) {
+        if (!isEmpty(caseData.getHearingDetails())) {
             hearingDate = " hearing " + dateFormatterService.formatLocalDateToString(
                 hearingBookingService.getMostUrgentHearingBooking(caseData.getHearingDetails()).getDate(),
                 FormatStyle.MEDIUM);
         }
         return Stream.of(subjectLine, hearingDate)
             .filter(StringUtils::isNotBlank)
-            .collect(joining(", "));
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: Was this change intentional? I'd imagine we'd want to have a space between our collections no?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java [hunk 2]
```diff
@@ -196,4 +204,71 @@ public static Document createDocumentWithAttachedFile() {
...
+    public static List<Element<C21Order>> createC21Orders() {
+        return ImmutableList.of(
+            Element.<C21Order>builder()
+                .value(C21Order.builder()
+                    .orderTitle("Example Order")
+                    .orderDetails(
+                        "Example order details here - Lorem ipsum dolor sit amet, consectetur adipiscing elit")
+                    .orderDate(DATE_FORMATTER_SERVICE.formatLocalDateTimeBaseUsingFormat(
+                        LocalDateTime.now().plusDays(57), FORMAT_STYLE))
+                    .judgeAndLegalAdvisor(JudgeAndLegalAdvisor.builder()
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java]: It might be of value splitting out building judgeAndLegalAdvisor into it's own method - `createJudgeAndLegalAdvisor`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java [hunk 3]
```diff
@@ -196,4 +204,71 @@ public static Document createDocumentWithAttachedFile() {
...
+                    .orderDetails("Westeros")
+                    .orderDate(DATE_FORMATTER_SERVICE.formatLocalDateTimeBaseUsingFormat(
+                        LocalDateTime.now().plusDays(59), FORMAT_STYLE))
+                    .judgeAndLegalAdvisor(JudgeAndLegalAdvisor.builder()
+                        .legalAdvisorName("Baratheon")
+                        .judgeFullName("Tyrion Lannister")
+                        .judgeLastName("Lannister")
+                        .judgeTitle(HIS_HONOUR_JUDGE)
+                        .build())
+                    .document(DocumentReference.builder()
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java]: Again might be of value splitting out into `createDocument`

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                                    randomUUID().toString(), "binary"))
+                                .build())
+                            .build())
+                        .build()),
+                "respondents1", createRespondents(),
+                "familyManCaseNumber", familyManCaseNumber))
+            .build();
+    }
+
+    private CaseDetails populateCaseDetails() {
```
Comment (Jamesnelson117) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: Not a great method name. Let's be more descriptive.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        return CaseDetails.builder()
+            .id(167888L)
+            .data(ImmutableMap.of("hearingDetails", createHearingBookings(LocalDate.now()),
+                "c21Orders", ImmutableList.of(
+                    Element.<C21Order>builder()
+                        .value(C21Order.builder()
+                            .orderTitle("Example Order")
+                            .orderDetails(
+                                "Example order details here - Lorem ipsum dolor sit amet, consectetur adipiscing elit")
+                            .judgeAndLegalAdvisor(JudgeAndLegalAdvisor.builder()
```
Comment (Jamesnelson117) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: If we split out our previous JudgeAndLegalAdvisor method into a separate static method, we could reuse here.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                                .legalAdvisorName("Peter Parker")
+                                .judgeLastName("Judy")
+                                .judgeTitle(HER_HONOUR_JUDGE)
+                                .build())
+                            .document(DocumentReference.builder()
```
Comment (Jamesnelson117) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: Re-use static method if you choose to extract.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java (hunk 1 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: yes it was as _hearingDate_ already had a space.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java [hunk 7]
```diff
@@ -43,24 +48,34 @@ protected C21OrderEmailContentProvider(@Value("${ccd.ui.base.url}")String uiBase
...
             "subjectLine", subjectLine,
             "localAuthorityOrCafcass", localAuthorityNameLookupConfiguration.getLocalAuthorityName(localAuthorityCode),
             "hearingDetailsCallout", buildSubjectLineWithHearingBookingDateSuffix(subjectLine, caseData),
+            "linkToDocStore", mostRecentUploadedC21DocumentUrl(caseData.getC21Orders()),
             "reference", String.valueOf(caseDetails.getId()),
             "caseUrl", uiBaseUrl + "/case/" + JURISDICTION + "/" + CASE_TYPE + "/" + caseDetails.getId()
         );
     }
 
+    private String mostRecentUploadedC21DocumentUrl(final List<Element<C21Order>> c21Orders) {
```
[the remark this answers is not in the document]
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: @Jamesnelson117 given we will be sending out the notification once the event occurs, will using the last upload not be preferred?

[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: Yep, ofc

File: service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java (hunk 2 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java]: good suggestion, updated

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java (hunk 4 above)
Comment (Gogetter) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: updated

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java [hunk 8]
```diff
@@ -43,24 +48,34 @@ protected C21OrderEmailContentProvider(@Value("${ccd.ui.base.url}")String uiBase
         // Validation within our frontend ensures that the following data is present
         CaseData caseData = objectMapper.convertValue(caseDetails.getData(), CaseData.class);
         final String subjectLine = buildSubjectLine(caseData);
-        return ImmutableMap.of(
+        return Map.of(
             "subjectLine", subjectLine,
             "localAuthorityOrCafcass", localAuthorityNameLookupConfiguration.getLocalAuthorityName(localAuthorityCode),
             "hearingDetailsCallout", buildSubjectLineWithHearingBookingDateSuffix(subjectLine, caseData),
+            "linkToDocStore", mostRecentUploadedC21DocumentUrl(caseData.getC21Orders()),
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: I don't like this implementation to be honest. The event should simply send the C21 Order which was attached, we shouldn't reach to the case and find out which one was added. 

Probably logic will be the same but in different place - event itself should be descriptive and know what happened.

Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: Why linkToDocStore and not simply linkToDocument? Just curious not 100% sure which one is better.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+import static uk.gov.hmcts.reform.fpl.utils.CaseDataGeneratorHelper.createDocumentReference;
+import static uk.gov.hmcts.reform.fpl.utils.CaseDataGeneratorHelper.createHearingBookings;
+import static uk.gov.hmcts.reform.fpl.utils.CaseDataGeneratorHelper.createJudgeAndLegalAdvisor;
+import static uk.gov.hmcts.reform.fpl.utils.CaseDataGeneratorHelper.createRespondents;
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = {JacksonAutoConfiguration.class, C21OrderEmailContentProvider.class,
+    HearingBookingService.class, LocalAuthorityNameLookupConfiguration.class, DateFormatterService.class})
+class C21OrderEmailContentProviderTest {
+    private final LocalDate date = LocalDate.now();
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: I would be more specific than "date", TODAY or something like that would make the tests more readable.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        assertThat(returnedParameters)
+            .extracting("subjectLine", "localAuthorityOrCafcass", "hearingDetailsCallout", "reference",
+                 "caseUrl")
+            .containsExactly(subjectLine, "Example Local Authority",
+                subjectLine + ", hearing " + dateFormatterService.formatLocalDateToString(date, FormatStyle.MEDIUM),
+                "167888", "/case/" + JURISDICTION + "/" + CASE_TYPE + "/167888");
+
+        /*  extracting separately as document_binary_url in populateCaseDetailsWithSingleC21Element()
+            method uses UUID.randomUUID()
+         */
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: Don't use randomUUID? Or pass externally the UUID value which should be used? If our test is checking not equal to empty on unit test level it's always smell to me. I change the implementation to always generate ABC and is it going to be correct implementation? Unit tests should spot that.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java (hunk 8 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: good suggestion on moving into the event itself in place of reaching into the _CaseData_. With regards the _linkToDocStore_, is it ok if I update the template on Gov dot notify and make this _linkToDocument_ as its more readable? @akomisarek

**Comment by Gogetter** [comment]
Hi @akomisarek, requested changes made. Please could you help take another look? thanks

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java [hunk 11]
```diff
@@ -38,71 +39,102 @@
...
+            cafcassLookupConfiguration);
 
         given(localAuthorityNameLookupConfiguration.getLocalAuthorityName(LOCAL_AUTHORITY_CODE))
             .willReturn("Example Local Authority");
 
+        given(cafcassLookupConfiguration.getCafcass(LOCAL_AUTHORITY_CODE))
+            .willReturn((new CafcassLookupConfiguration.Cafcass(CAFCASS_NAME, CAFCASS_EMAIL)));
+
         familyManCaseNumber = RandomStringUtils.randomAlphabetic(8);
+        documentId = randomUUID().toString();
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: Really picky, but if we have got documentId I believe it should be in UUID format and defer toString when it's required.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java [hunk 12]
```diff
@@ -38,71 +39,102 @@
...
-        /*  extracting separately as document_binary_url in populateCaseDetailsWithSingleC21Element()
-            method uses UUID.randomUUID()
-         */
-        assertThat(returnedParameters).extracting("linkToDocStore").isNotEqualTo("");
+        assertThat(returnedLocalAuthorityParameters)
+            .extracting("subjectLine", "localAuthorityOrCafcass", "hearingDetailsCallout",
+                "linkToDocument", "reference", "caseUrl")
+            .containsExactly(subjectLine, "Example Local Authority",
+                (subjectLine + ", hearing " + dateFormatterService.formatLocalDateToString(today, FormatStyle.MEDIUM)),
+                documentUrl, "167888", "/case/" + JURISDICTION + "/" + CASE_TYPE + "/167888");
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: Isn't this another test as we are checking something else? The unit test could assert one thing.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java [hunk 13]
```diff
@@ -38,71 +39,102 @@
...
-
-        Map<String, Object> returnedParameters = c21OrderEmailContentProvider.buildC21OrderNotification(caseDetails,
-            LOCAL_AUTHORITY_CODE);
-
-        final String subjectLine = "Jones, " + familyManCaseNumber;
-
-        assertThat(returnedParameters)
-            .extracting("subjectLine", "localAuthorityOrCafcass", "hearingDetailsCallout", "linkToDocStore",
-                "reference", "caseUrl")
+        final String mostRecentDocumentUrl = "http://dm-store:8080/documents/79ec80ec-7be6-493b-b4e6-f002f05b7079/binary";
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: Similar here, we are checking multiple scenarios.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java [hunk 14]
```diff
@@ -96,4 +106,13 @@ private String getCourtName(String courtName) {
                     .formatLocalDateToString(child.getDateOfBirth(), FormatStyle.LONG) : ""))
             .collect(toList());
     }
+
+    private String mostRecentUploadedC21DocumentUrl(final List<Element<C21Order>> c21Orders) {
+        return getLast(c21Orders.stream()
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java]: Hmm I am not sure if I like the approach here, getLast relies on the order provided to us by CCD, I think it's bad practice and makes code a lot more difficult in future to change than required. I know it's probably correct in this instance (and it will be 99% of the time in future), but maybe we should return really what mostRecent means? order by date and take the latest one?

Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java]: Probbaly before collecting mapping would be a bit nicer :)

Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java]: @joez-f what do you think about the above?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderControllerTest.java [hunk 15]
```diff
@@ -137,4 +176,58 @@ private AboutToStartOrSubmitCallbackResponse makeRequest(CallbackRequest request
...
+                    "c21Orders", createC21Orders(),
+                    "hearingDetails", createHearingBookings(dateIn3Months),
+                    "respondents1", createRespondents(),
+                    "caseLocalAuthority", LOCAL_AUTHORITY_CODE,
+                    "familyManCaseNumber", FAMILY_MAN_CASE_NUMBER))
+                .build())
+            .build();
+    }
+
+    private Map<String, Object> similarParameters() {
```
Comment (akomisarek) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderControllerTest.java]: Not sure if I like the name similar, similar to what? commonNotification?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java (hunk 14 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java]: I like sorting by orderDate and then taking the latest one - good to do it explicitly (although I do agree that in this instance I don't see a way that the list of c21Orders could ever be out of order. so not a huge problem IMO, but let's explicitly return mostRecent if it's not too difficult)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/events/C21OrderEvent.java [hunk 16]
```diff
@@ -1,10 +1,25 @@
...
+
+    public C21OrderEvent(CallbackRequest callbackRequest, String authorization, String userId,
+                         C21OrderEventData c21OrderEventData) {
         super(callbackRequest, authorization, userId);
+        this.c21OrderEventData = c21OrderEventData;
+    }
+
+    @Value
+    @Builder
+    public static class C21OrderEventData {
```
[the remark this answers is not in the document]
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/events/C21OrderEvent.java]: yes was thinking that myself after writing. The idea was to split the document link out. But given its only one field we can pass as a param

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java (hunk 14 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java]: I picked this option as the _orderDate_ is currently stored in string

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderControllerTest.java (hunk 15 above)
Comment (Gogetter) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderControllerTest.java]: updated

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderControllerTest.java [hunk 17]
```diff
@@ -219,14 +219,14 @@ private CallbackRequest buildCallbackRequest() {
 
     private Map<String, Object> c21CafcassParameters() {
```
Comment (akomisarek) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderControllerTest.java]: Wouldn't it be beneficial to name them "expected" that looks like they are used in beforeEach and used for setup.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/config/DocumentConfiguration.java [hunk 18]
```diff
@@ -0,0 +1,15 @@
+package uk.gov.hmcts.reform.fpl.config;
+
+import lombok.Data;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Configuration;
+
+@Data
+@Configuration
+public class DocumentConfiguration {
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/config/DocumentConfiguration.java]: I don't like the name DocumentConfiguration as it doesn't have to do anything with the document - it's the GatewayConfiguration and are configuring the gatewayUrl. 

At least DocumentGateway? but still the use is more broader than just documents.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/config/DocumentConfiguration.java [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+    private final String documentHostUrl;
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/config/DocumentConfiguration.java]: As above.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java [hunk 20]
```diff
@@ -107,12 +107,22 @@ private String getCourtName(String courtName) {
             .collect(toList());
     }
 
-    private String mostRecentUploadedC21DocumentUrl(final List<Element<C21Order>> c21Orders) {
-        return getLast(c21Orders.stream()
+    public String mostRecentUploadedC21DocumentUrl(final List<Element<C21Order>> c21Orders) {
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java]: Please implement unit test for this method now.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java [hunk 21]
```diff
@@ -107,12 +107,22 @@ private String getCourtName(String courtName) {
...
+
+        final String documentHostUrl = documentConfiguration.getDocumentHostUrl();
+
+        try {
+            URL url = new URL(mostRecentUploadedC21DocumentUrl);
+            mostRecentUploadedC21DocumentUrl = documentHostUrl + url.getPath();
+        } catch (MalformedURLException e) {
+            log.error("mostRecentUploadedC21DocumentUrl url incorrect.", e);
+        }
+        return mostRecentUploadedC21DocumentUrl;
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderService.java]: Could we extract to sdeparatee method? A bit messy. Could allow some mapping. 

Can you swap from URL to URI? UIR is a bit more lightweight class. 

Not sure about returning original URL if it fail to parse, probably would be better to return some placeholder text? It would be invalid URL so not sure.
Are you sure the "e" contains the URL for the debugging purposes?

File: service/src/main/resources/application.yaml [hunk 22]
```diff
@@ -104,6 +105,7 @@ docmosis:
 
 document_management:
   url: http://localhost:3453
+  dm_host_url: https://gateway-${SERVICE_FQDN}
```
Comment (akomisarek) [inline service/src/main/resources/application.yaml]: This cannot contain variable + it's not dm host.

File: service/src/main/resources/application.yaml [hunk 23]
```diff
@@ -44,6 +44,7 @@ docmosis:
 
 document_management:
   url: http://fake-document-management-store-api
+  dm_host_url: http://fake-document-gateway-host
```
Comment (akomisarek) [inline service/src/main/resources/application.yaml]: It's not DM host.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java [hunk 24]
```diff
@@ -118,7 +118,7 @@
 
     @Nested
     class C2UploadedNotificationChecks {
-        final String documentUrl = "http://dm-store:8080/documents/79ec80ec-7be6-493b-b4e6-f002f05b7079/binary";
+        final String mostRecentUploadedDocumentUrl = "http://dm-store:8080/documents/79ec80ec-7be6-493b-b4e6-f002f05b7079/binary";
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java]: Really picky but it would be better if we do gateway URL here to avoid confusion.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderServiceTest.java [hunk 25]
```diff
@@ -39,12 +40,14 @@
 
     private DateFormatterService dateFormatterService = new DateFormatterService();
     private HmctsCourtLookupConfiguration hmctsCourtLookupConfiguration = new HmctsCourtLookupConfiguration(CONFIG);
+    private DocumentConfiguration documentConfiguration = new DocumentConfiguration("");
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/CreateC21OrderServiceTest.java]: Not sure why this needs to be field if neverused in class (apart from beforeEach)

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java [hunk 26]
```diff
@@ -139,6 +157,24 @@ void shouldReturnExactC21NotificationParametersWithMostRecentUploadedDocumentUrl
                 "/case/" + JURISDICTION + "/" + CASE_TYPE + "/12345");
     }
 
+    @Test
+    void shouldReturnExactC21LocalAuthorityNotificationParametersWithMostRecentUploadedDocumentUrl() {
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: What is different in this test from the one declared above in line 115? Multi c21 are not relevant on that level as the URL is just provided in the contentProvider so why we need to test separate scenario? Is it even useful?

File: charts/fpl-case-service/values.aat.template.yaml [hunk 27]
```diff
@@ -13,6 +13,7 @@ java:
     SPRING_SECURITY_ENABLED: true
     SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_ISSUERURI: https://forgerock-am.service.core-compute-idam-aat.internal:8443/openam/oauth2/hmcts
     SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_JWKSETURI: https://idam-api.aat.platform.hmcts.net/o/jwks
+    DOCUMENT_MANAGEMENT_GATEWAY_URL: https://gateway-${SERVICE_FQDN}
```
Comment (akomisarek) [inline charts/fpl-case-service/values.aat.template.yaml]: Not sure why you insist of using the document management, I am not sure if we should say it's document management, as this is just generic gateway? Different purpose?

File: charts/fpl-case-service/values.preview.template.yaml [hunk 28]
```diff
@@ -13,6 +13,7 @@ java:
     SPRING_SECURITY_ENABLED: true
     SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_ISSUERURI: https://forgerock-am.service.core-compute-idam-aat.internal:8443/openam/oauth2/hmcts
     SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_JWKSETURI: https://idam-api.aat.platform.hmcts.net/o/jwks
+    DOCUMENT_MANAGEMENT_GATEWAY_URL: https://gateway-${SERVICE_FQDN}
```
Comment (akomisarek) [inline charts/fpl-case-service/values.preview.template.yaml]: doesn't we need to have -ccd? 
The above will result in
gateway-fpl-case-service-pr-618.service.core-compute-preview.internal. Is this correct URL? Can you access it?

File: infrastructure/prod.tfvars [hunk 29]
```diff
@@ -9,3 +9,4 @@ docmosis_vault = "docmosisiaasprodkv"
 
 idam_client_id = "fpl_case_service"
 idam_client_redirect_uri = "https://fpl-case-service-prod.service.core-compute-prod.internal/oauth2/callback"
+document_management_gateway_url = "https://gateway-ccd.prod.platform.hmcts.net"
```
Comment (akomisarek) [inline infrastructure/prod.tfvars]: Apart from that you need to configure it for the demo (which is only real environment using the terraform to that extend)

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderControllerTest.java [hunk 30]
```diff
@@ -155,11 +155,11 @@ void shouldTriggerC21EventWhenSubmitted() throws Exception {
 
         verify(notificationClient, times(1)).sendEmail(
             eq(C21_ORDER_NOTIFICATION_TEMPLATE), eq(LOCAL_AUTHORITY_EMAIL_ADDRESS),
-            eq(c21LocalAuthorityParameters()), eq(expectedCaseReference));
+            eq(expectedC21LocalAuthorityParameters()), eq(expectedCaseReference));
```
Comment (akomisarek) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderControllerTest.java]: 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/config/GatewayConfiguration.java [hunk 31]
```diff
@@ -0,0 +1,15 @@
+package uk.gov.hmcts.reform.fpl.config;
+
+import lombok.Data;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Configuration;
+
+@Data
+@Configuration
+public class GatewayConfiguration {
+    private final String gatewayUrl;
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/config/GatewayConfiguration.java]: Isn't it enough to name it url? it's obvious its gateway as we have it in the Class name :)

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java (hunk 26 above)
Comment (Gogetter) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: yes looks very similar. Line 115 bit checks for Cafcass parameters and this checks for LA params

File: service/src/main/java/uk/gov/hmcts/reform/fpl/config/GatewayConfiguration.java [hunk 32]
```diff
@@ -0,0 +1,15 @@
...
+
+    public GatewayConfiguration(@Value("${document_management.gateway_url}") String gatewayUrl) {
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/config/GatewayConfiguration.java]: Maybe gateway? And not document management, but not that fussed if you really like the doc_management prefix.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderController.java [hunk 33]
```diff
@@ -121,4 +129,17 @@ private Document getDocument(String authorization,
...
+        final String gatewayUrl = gatewayConfiguration.getGatewayUrl();
+
+        try {
+            URI uri = new URI(mostRecentUploadedC21DocumentUrl);
+            return gatewayUrl + uri.getPath();
+        } catch (URISyntaxException e) {
+            log.error(mostRecentUploadedC21DocumentUrl + " url incorrect.", e);
+        }
+        return "";
+    }
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderController.java]: I am not sure if the controller is however the good place for this logic? maybe could be done in notification content provider? This is the place which really needs to transform the URL, on the controller level it's ok, I think?

Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderController.java]: Would be easier to test as well - and this piece of code should be unit tested.

File: charts/fpl-case-service/values.aat.template.yaml (hunk 27 above)
Comment (Gogetter) [reply charts/fpl-case-service/values.aat.template.yaml]: for now was trying to make it as specific as possible. But you have made a good

Comment (Gogetter) [reply charts/fpl-case-service/values.aat.template.yaml]: Lets keep as gateway given the class name 👍

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java [hunk 34]
```diff
@@ -0,0 +1,184 @@
...
+                "linkToDocument", "reference", "caseUrl")
+            .containsExactly(subjectLine, "Example Cafcass",
+                subjectLine + ", hearing " + dateFormatterService.formatLocalDateToString(today, FormatStyle.MEDIUM),
+                 mostRecentDocumentUrl, "12345", "/case/" + JURISDICTION + "/" + CASE_TYPE + "/12345");
+    }
+
+    @Test
+    void shouldReturnExactC21LocalAuthorityNotificationParametersWithMostRecentUploadedDocumentUrl() {
+        final String mostRecentDocumentUrl = "http://dm-store:8080/documents/79ec80ec-7be6-493b-b4e6-f002f05b7079/binary";
+        CaseDetails caseDetails = createCaseDetailsWithMultipleC21Elements();
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/C21OrderEmailContentProviderTest.java]: As agreed this test should be removed.
